Method Single::chunkwise() added

Smoren opened the pull request:
1. `Single::chunkwise()` method added.
2. Tests added for new method.
3. README updated with description of new method.

Smoren committed:
1. `Single::chunkwise()` method added.
2. Tests added for new method.
3. README updated with description of new method.

Smoren committed:
`Single::chunkwise()`: error test added.

Smoren committed:
1. `Stream::chunkwise()` method added.
2. Tests added for new method.
3. README updated with description of new method.
4. Some bugs fixed in README.

Smoren committed:
`Single::chunkwiseOverlap()` method added.

markrogoyski commented:
Hi @Smoren,

Thanks for the PR. Chunking makes a lot of sense and was on my roadmap of things to implement. Let me know when you are finished and I'll merge it in.

Thanks again,
Mark

Smoren committed:
Methods `Single::pairwise()` and `Single::chunkwise()` refactored with using method `Single::chunkwiseOverlap()`.

Smoren committed:
`Single::chunkwiseOverlap()`: PHPdoc and README updated.

Smoren committed:
1. `Stream::chunkwiseOverlap()` method added.
2. Tests added for new method.
3. README updated with description of new method.

Smoren commented:
Hi @markrogoyski

I think I've finished this branch.

Summary:

1. Methods added:
    * `Single::chunkwise(iterable $data, int $chunkSize)`,
    * `Single::chunkwiseOverlap(iterable $data, int $chunkSize, int $overlapSize)`,
    * `Stream::chunkwise(int $chunkSize): self`,
    * `Stream::chunkwiseOverlap(int $chunkSize, int $overlapSize): self`.
2. Methods `Single::pairwise()` and `Single::chunkwise()` are refactored using `Single::chunkwiseOverlap()`.
3. Tests added for all the new methods.
4. README updated with new methods' descriptions.

FYI: Strange thing. When I run coverage it says that the `Stream::infiniteCycle()` is not covered by tests. But I also see that it is really covered.

![image](https://user-images.githubusercontent.com/7403235/212495837-bc556886-aed0-458b-9ccf-b1f9bf5622cf.png)

Smoren committed:
phpstan and codesniffer fixes

Smoren commented:
@markrogoyski 

I've rebased this branch from the `develop` and fixed codesniffer and phpstan issues.

Now I think PR is ready.

markrogoyski commented on tests/Single/ChunkwiseTest.php:
```diff
@@ -0,0 +1,651 @@
+     * @param iterable $data
+     * @param int $chunkSize
+     */
+    public function testError(iterable $data, int $chunkSize): void
+    {
+        // Given
+        try {
+            // When
+            foreach (Single::chunkwise($data, $chunkSize) as $_) {
+                break;
+            }
+            $this->fail();
+        } catch (\InvalidArgumentException $e) {
+            // Then
+            $this->assertEquals("Chunk size must be ≥ 1. Got {$chunkSize}", $e->getMessage());
+        }
+    }
```
FYI: You can simplify this test by using PHPUnit's exception assertion mechanism and have a flat test that is much simpler.

```php
        // Then
        $this->expectException(\InvalidArgumentException::class);
        $this->expectExceptionMessage("Chunk size must be ≥ 1. Got {$chunkSize}");

        // When
        foreach (Single::chunkwise($data, $chunkSize) as $_) {
            break;
        }
```

The test will fail if no exception is thrown. I've updated the tests in develop.

Smoren replied:
Thanks for the helpful advice!

Smoren commented:
@markrogoyski 

Thank you for merging this PR!

By the way, did you find the cause of the strange thing about the coverage that I mentioned in one of the previous comments?

markrogoyski commented:
Hi @Smoren,

I think the code coverage may have to do with whether the method is called during the body of a unit test, or during the setup. I think I actually changed from using the lazy anonymous function to calling it in the data provider. Adding a single test that calls the method in the test should probably add it to the code coverage.

Mark

Smoren commented:
@markrogoyski 

Thanks. Please show me the decision when you apply it.

markrogoyski commented:
Hi @Smoren,

This is the commit where I refactored the unit tests from a lazy anonymous function to directly creating the stream in the unit test data provider: https://github.com/markrogoyski/itertools-php/commit/e9472959b84bf54086d5599b67aadbc67f997694  Doing this removed those lines of code from being considered for the code coverage calculation.

I just pushed a commit to develop that adds a new unit test in the style you originally wrote that will defer executing the stream until the unit test runs: https://github.com/markrogoyski/itertools-php/commit/47cb65a8cf74ec3f2b1f93bd354ea3890aae2bd5

And the updated code coverage confirms it is fixed:
![infinitecycle code coverage](https://user-images.githubusercontent.com/10004372/212560319-144ee439-7f75-4ef9-9337-30964f6ccbfb.png)

This has to due with PHPUnit running all the data provider prior to executing any tests, so it doesn't factor in code run in the data providers as part of the covering tests.

Smoren commented:
Thank you @markrogoyski
